File: src/main/kotlin/org/skialin/Canvas.kt
package org.skialin

import org.skialin.impl.NativeLoader

class Canvas internal constructor(
    internal val ptr: Long,
) {
    fun clear(color: Color) = CanvasNative.nClear(ptr, color)

    fun drawColor(
        color: Color,
        mode: BlendMode = BlendMode.SRC_OVER,
    ) = CanvasNative.nDrawColor(ptr, color, mode.ordinal)

    fun drawPaint(paint: Paint) = CanvasNative.nDrawPaint(ptr, paint.nativePtr)

    fun drawLine(
        p0: Point,
        p1: Point,
        paint: Paint,
    ) = CanvasNative.nDrawLine(ptr, p0.x, p0.y, p1.x, p1.y, paint.nativePtr)

    fun drawRect(
        rect: Rect,
        paint: Paint,
    ) = CanvasNative.nDrawRect(ptr, rect.left, rect.top, rect.right, rect.bottom, paint.nativePtr)

    fun drawOval(
        rect: Rect,
        paint: Paint,
    ) = CanvasNative.nDrawOval(ptr, rect.left, rect.top, rect.right, rect.bottom, paint.nativePtr)

    fun drawCircle(
        center: Point,
        radius: Float,
        paint: Paint,
    ) = CanvasNative.nDrawCircle(ptr, center.x, center.y, radius, paint.nativePtr)

    fun drawPath(
        path: Path,
        paint: Paint,
    ) = CanvasNative.nDrawPath(ptr, path.nativePtr, paint.nativePtr)

    fun drawTextBlob(
        blob: TextBlob,
        x: Float,
        y: Float,
        paint: Paint,
    ) = CanvasNative.nDrawTextBlob(ptr, blob.nativePtr, x, y, paint.nativePtr)

    fun save(): Int = CanvasNative.nSave(ptr)

    fun restore() = CanvasNative.nRestore(ptr)

    fun restoreToCount(saveCount: Int) = CanvasNative.nRestoreToCount(ptr, saveCount)

    fun translate(
        dx: Float,
        dy: Float,
    ) = CanvasNative.nTranslate(ptr, dx, dy)

    fun scale(
        sx: Float,
        sy: Float,
    ) = CanvasNative.nScale(ptr, sx, sy)

    fun rotate(degrees: Float) = CanvasNative.nRotate(ptr, degrees)

    fun clipRect(
        rect: Rect,
        op: ClipOp = ClipOp.INTERSECT,
    ) = CanvasNative.nClipRect(ptr, rect.left, rect.top, rect.right, rect.bottom, op.ordinal)

    fun clipPath(
        path: Path,
        op: ClipOp = ClipOp.INTERSECT,
    ) = CanvasNative.nClipPath(ptr, path.nativePtr, op.ordinal)

    fun skew(
        sx: Float,
        sy: Float,
    ) = CanvasNative.nSkew(ptr, sx, sy)

    fun resetMatrix() = CanvasNative.nResetMatrix(ptr)

    fun setMatrix(matrix: Matrix33) = CanvasNative.nSetMatrix(ptr, matrix.values)

    fun getTotalMatrix(): Matrix33 = Matrix33(CanvasNative.nTotalMatrix(ptr))

    fun quickReject(rect: Rect): Boolean = CanvasNative.nQuickRejectRect(ptr, rect.left, rect.top, rect.right, rect.bottom)

    fun quickReject(path: Path): Boolean = CanvasNative.nQuickRejectPath(ptr, path.nativePtr)

    fun drawRoundRect(
        rect: Rect,
        rx: Float,
        ry: Float,
        paint: Paint,
    ) = CanvasNative.nDrawRoundRect(ptr, rect.left, rect.top, rect.right, rect.bottom, rx, ry, paint.nativePtr)

    fun drawArc(
        oval: Rect,
        startAngle: Float,
        sweepAngle: Float,
        useCenter: Boolean,
        paint: Paint,
    ) = CanvasNative.nDrawArc(ptr, oval.left, oval.top, oval.right, oval.bottom, startAngle, sweepAngle, useCenter, paint.nativePtr)

    fun drawPoints(
        mode: PointMode,
        points: Array<Point>,
        paint: Paint,
    ) {
        val flat = FloatArray(points.size * 2)
        points.forEachIndexed { i, p ->
            flat[i * 2] = p.x
            flat[i * 2 + 1] = p.y
        }
        CanvasNative.nDrawPoints(ptr, mode.ordinal, flat, paint.nativePtr)
    }

    fun drawImage(
        image: Image,
        x: Float,
        y: Float,
        sampling: SamplingOptions = SamplingOptions.NEAREST,
        paint: Paint? = null,
    ) = CanvasNative.nDrawImage(
        ptr,
        image.nativePtr,
        x,
        y,
        sampling.maxAniso,
        sampling.useCubic,
        sampling.cubicB ?: 0f,
        sampling.cubicC ?: 0f,
        sampling.filter.ordinal,
        sampling.mipmap.ordinal,
        paint?.nativePtr ?: 0L,
    )

    /** [src] defaults to the whole image when `null`. */
    fun drawImageRect(
        image: Image,
        dst: Rect,
        src: Rect? = null,
        sampling: SamplingOptions = SamplingOptions.NEAREST,
        paint: Paint? = null,
        constraint: SrcRectConstraint = SrcRectConstraint.STRICT,
    ) = CanvasNative.nDrawImageRect(
        ptr,
        image.nativePtr,
        src?.let { floatArrayOf(it.left, it.top, it.right, it.bottom) },
        dst.left,
        dst.top,
        dst.right,
        dst.bottom,
        sampling.maxAniso,
        sampling.useCubic,
        sampling.cubicB ?: 0f,
        sampling.cubicC ?: 0f,
        sampling.filter.ordinal,
        sampling.mipmap.ordinal,
        paint?.nativePtr ?: 0L,
        constraint.ordinal,
    )

    /** Draws [image] as a 9-patch: [center] scales to fill the interior of [dst], while the
     * surrounding edges/corners scale only along one axis (or not at all). */
    fun drawImageNine(
        image: Image,
        center: IRect,
        dst: Rect,
        filter: FilterMode = FilterMode.NEAREST,
        paint: Paint? = null,
    ) = CanvasNative.nDrawImageNine(
        ptr,
        image.nativePtr,
        intArrayOf(center.left, center.top, center.right, center.bottom),
        dst.left,
        dst.top,
        dst.right,
        dst.bottom,
        filter.ordinal,
        paint?.nativePtr ?: 0L,
    )

    /** Draws a Coons patch: [cubics] is the 12-point boundary (4 cubic Bezier edges sharing
     * corner points), [colors] are the 4 corner colors, [texCoords] optionally maps a source
     * image's 4 corners onto the patch via a shader on [paint]. */
    fun drawPatch(
        cubics: Array<Point>,
        colors: Array<Color>,
        texCoords: Array<Point>? = null,
        mode: BlendMode = BlendMode.SRC_OVER,
        paint: Paint,
    ) {
        require(cubics.size == 12) { "cubics must have 12 points" }
        require(colors.size == 4) { "colors must have 4 entries" }
        require(texCoords == null || texCoords.size == 4) { "texCoords must have 4 points" }
        val flatCubics = FloatArray(24)
        cubics.forEachIndexed { i, p ->
            flatCubics[i * 2] = p.x
            flatCubics[i * 2 + 1] = p.y
        }
        val flatTex =
            texCoords?.let { coords ->
                FloatArray(8).also { flat ->
                    coords.forEachIndexed { i, p ->
                        flat[i * 2] = p.x
                        flat[i * 2 + 1] = p.y
                    }
                }
            }
        CanvasNative.nDrawPatch(ptr, flatCubics, colors.toIntArray(), flatTex, mode.ordinal, paint.nativePtr)
    }

    /** Attaches a key/value annotation to the document at [rect] (e.g. a hyperlink or named
     * destination when drawing into a PDF-backed canvas); a no-op for raster/GPU canvases. */
    fun drawAnnotation(
        rect: Rect,
        key: String,
        value: Data? = null,
    ) = CanvasNative.nDrawAnnotation(ptr, rect.left, rect.top, rect.right, rect.bottom, key, value?.nativePtr ?: 0L)

    fun drawRRect(
        rrect: RRect,
        paint: Paint,
    ) = CanvasNative.nDrawRRect(ptr, rrect.nativePtr, paint.nativePtr)

    fun drawPicture(picture: Picture) = CanvasNative.nDrawPicture(ptr, picture.nativePtr)

    /** Draws the ring between [outer] and [inner]; [inner] must be contained within [outer]. */
    fun drawDRRect(
        outer: RRect,
        inner: RRect,
        paint: Paint,
    ) = CanvasNative.nDrawDRRect(ptr, outer.nativePtr, inner.nativePtr, paint.nativePtr)

    fun clipRRect(
        rrect: RRect,
        op: ClipOp = ClipOp.INTERSECT,
    ) = CanvasNative.nClipRRect(ptr, rrect.nativePtr, op.ordinal)

    fun drawRegion(
        region: Region,
        paint: Paint,
    ) = CanvasNative.nDrawRegion(ptr, region.nativePtr, paint.nativePtr)

    fun clipRegion(
        region: Region,
        op: ClipOp = ClipOp.INTERSECT,
    ) = CanvasNative.nClipRegion(ptr, region.nativePtr, op.ordinal)

    fun drawVertices(
        vertices: Vertices,
        mode: BlendMode,
        paint: Paint,
    ) = CanvasNative.nDrawVertices(ptr, vertices.nativePtr, mode.ordinal, paint.nativePtr)

    /** Concatenates a 4x4 local-to-device transform onto the canvas's current matrix. */
    fun concat44(matrix: M44) = CanvasNative.nConcat44(ptr, matrix.nativePtr)

    /**
     * Saves the canvas state, then redirects drawing to a new layer. [bounds], if given, is
     * a hint for the layer's extent. Returns the new save count, for [restoreToCount].
     */
    fun saveLayer(
        bounds: Rect? = null,
        paint: Paint? = null,
    ): Int = CanvasNative.nSaveLayer(ptr, bounds?.let { floatArrayOf(it.left, it.top, it.right, it.bottom) }, paint?.nativePtr ?: 0L)

    /**
     * Like [saveLayer], but [backdrop], if given, filters the current layer's content before
     * it's drawn into the new one (instead of the new layer starting out transparent-black).
     */
    fun saveLayer(
        bounds: Rect? = null,
        paint: Paint? = null,
        backdrop: ImageFilter?,
        flags: Int = 0,
    ): Int =
        CanvasNative.nSaveLayerWithBackdrop(
            ptr,
            bounds?.let { floatArrayOf(it.left, it.top, it.right, it.bottom) },
            paint?.nativePtr ?: 0L,
            backdrop?.nativePtr ?: 0L,
            flags,
        )

    /** Runs [block] between [save] and [restore]. */
    inline fun withSave(block: Canvas.() -> Unit) {
        save()
        try {
            block()
        } finally {
            restore()
        }
    }
}

private object CanvasNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nClear(
        ptr: Long,
        color: Int,
    )

    external fun nDrawColor(
        ptr: Long,
        color: Int,
        mode: Int,
    )

    external fun nDrawPaint(
        ptr: Long,
        paintPtr: Long,
    )

    external fun nDrawLine(
        ptr: Long,
        x0: Float,
        y0: Float,
        x1: Float,
        y1: Float,
        paintPtr: Long,
    )

    external fun nDrawRect(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        paintPtr: Long,
    )

    external fun nDrawOval(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        paintPtr: Long,
    )

    external fun nDrawCircle(
        ptr: Long,
        cx: Float,
        cy: Float,
        radius: Float,
        paintPtr: Long,
    )

    external fun nDrawPath(
        ptr: Long,
        pathPtr: Long,
        paintPtr: Long,
    )

    external fun nDrawTextBlob(
        ptr: Long,
        blobPtr: Long,
        x: Float,
        y: Float,
        paintPtr: Long,
    )

    external fun nSave(ptr: Long): Int

    external fun nRestore(ptr: Long)

    external fun nRestoreToCount(
        ptr: Long,
        saveCount: Int,
    )

    external fun nTranslate(
        ptr: Long,
        dx: Float,
        dy: Float,
    )

    external fun nScale(
        ptr: Long,
        sx: Float,
        sy: Float,
    )

    external fun nRotate(
        ptr: Long,
        degrees: Float,
    )

    external fun nClipRect(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        op: Int,
    )

    external fun nClipPath(
        ptr: Long,
        pathPtr: Long,
        op: Int,
    )

    external fun nSkew(
        ptr: Long,
        sx: Float,
        sy: Float,
    )

    external fun nResetMatrix(ptr: Long)

    external fun nSetMatrix(
        ptr: Long,
        matrix: FloatArray,
    )

    external fun nTotalMatrix(ptr: Long): FloatArray

    external fun nQuickRejectRect(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
    ): Boolean

    external fun nQuickRejectPath(
        ptr: Long,
        pathPtr: Long,
    ): Boolean

    external fun nDrawRoundRect(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        rx: Float,
        ry: Float,
        paintPtr: Long,
    )

    external fun nDrawArc(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        startAngle: Float,
        sweepAngle: Float,
        useCenter: Boolean,
        paintPtr: Long,
    )

    external fun nDrawPoints(
        ptr: Long,
        mode: Int,
        points: FloatArray,
        paintPtr: Long,
    )

    external fun nDrawImage(
        ptr: Long,
        imagePtr: Long,
        x: Float,
        y: Float,
        maxAniso: Int,
        useCubic: Boolean,
        cubicB: Float,
        cubicC: Float,
        filter: Int,
        mipmap: Int,
        paintPtr: Long,
    )

    external fun nDrawImageRect(
        ptr: Long,
        imagePtr: Long,
        src: FloatArray?,
        dstLeft: Float,
        dstTop: Float,
        dstRight: Float,
        dstBottom: Float,
        maxAniso: Int,
        useCubic: Boolean,
        cubicB: Float,
        cubicC: Float,
        filter: Int,
        mipmap: Int,
        paintPtr: Long,
        constraint: Int,
    )

    external fun nSaveLayer(
        ptr: Long,
        bounds: FloatArray?,
        paintPtr: Long,
    ): Int

    external fun nSaveLayerWithBackdrop(
        ptr: Long,
        bounds: FloatArray?,
        paintPtr: Long,
        backdropPtr: Long,
        flags: Int,
    ): Int

    external fun nDrawImageNine(
        ptr: Long,
        imagePtr: Long,
        center: IntArray,
        dstLeft: Float,
        dstTop: Float,
        dstRight: Float,
        dstBottom: Float,
        filter: Int,
        paintPtr: Long,
    )

    external fun nDrawPatch(
        ptr: Long,
        cubics: FloatArray,
        colors: IntArray,
        texCoords: FloatArray?,
        mode: Int,
        paintPtr: Long,
    )

    external fun nDrawAnnotation(
        ptr: Long,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
        key: String,
        valuePtr: Long,
    )

    external fun nDrawRRect(
        ptr: Long,
        rrectPtr: Long,
        paintPtr: Long,
    )

    external fun nDrawPicture(
        ptr: Long,
        picturePtr: Long,
    )

    external fun nDrawDRRect(
        ptr: Long,
        outerPtr: Long,
        innerPtr: Long,
        paintPtr: Long,
    )

    external fun nClipRRect(
        ptr: Long,
        rrectPtr: Long,
        op: Int,
    )

    external fun nDrawRegion(
        ptr: Long,
        regionPtr: Long,
        paintPtr: Long,
    )

    external fun nClipRegion(
        ptr: Long,
        regionPtr: Long,
        op: Int,
    )

    external fun nDrawVertices(
        ptr: Long,
        verticesPtr: Long,
        mode: Int,
        paintPtr: Long,
    )

    external fun nConcat44(
        ptr: Long,
        matrixPtr: Long,
    )
}

File: rust/skialin-core/src/canvas.rs
use std::marker::PhantomData;

use crate::paint::BlendMode;
use crate::path::Path;
use crate::{sys, Bitmap, Color, Data, FilterMode, Font, Image, IRect, Matrix, Paint, Picture, Point, RRect, Rect, Region, SamplingOptions, TextBlob, Vertices, M44};

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum PointMode {
    Points,
    Lines,
    Polygon,
}

impl From<PointMode> for sys::SkCanvas_PointMode {
    fn from(mode: PointMode) -> Self {
        (match mode {
            PointMode::Points => sys::SkCanvas_PointMode_kPoints_PointMode,
            PointMode::Lines => sys::SkCanvas_PointMode_kLines_PointMode,
            PointMode::Polygon => sys::SkCanvas_PointMode_kPolygon_PointMode,
        }) as sys::SkCanvas_PointMode
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum SrcRectConstraint {
    Strict,
    Fast,
}

impl From<SrcRectConstraint> for sys::SkCanvas_SrcRectConstraint {
    fn from(constraint: SrcRectConstraint) -> Self {
        (match constraint {
            SrcRectConstraint::Strict => sys::SkCanvas_SrcRectConstraint_kStrict_SrcRectConstraint,
            SrcRectConstraint::Fast => sys::SkCanvas_SrcRectConstraint_kFast_SrcRectConstraint,
        }) as sys::SkCanvas_SrcRectConstraint
    }
}

pub(crate) fn to_sk_sampling(sampling: SamplingOptions) -> sys::SkSamplingOptions {
    let (cubic_b, cubic_c) = sampling.cubic.unwrap_or((0.0, 0.0));
    sys::SkSamplingOptions {
        maxAniso: sampling.max_aniso,
        useCubic: sampling.cubic.is_some(),
        cubic: sys::SkCubicResampler { B: cubic_b, C: cubic_c },
        filter: sampling.filter.into(),
        mipmap: sampling.mipmap.into(),
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum ClipOp {
    Difference,
    Intersect,
}

impl From<ClipOp> for sys::SkClipOp {
    fn from(op: ClipOp) -> Self {
        (match op {
            ClipOp::Difference => 0,
            ClipOp::Intersect => 1,
        }) as sys::SkClipOp
    }
}

pub struct Canvas<'a> {
    pub(crate) ptr: *mut sys::SkCanvas,
    pub(crate) _marker: PhantomData<&'a mut ()>,
}

impl<'a> Canvas<'a> {
    pub unsafe fn from_raw(ptr: *mut sys::SkCanvas) -> Self {
        Canvas { ptr, _marker: PhantomData }
    }

    pub fn new_from_bitmap(bitmap: &mut Bitmap) -> Self {
        let ptr = unsafe { sys::skialin_bridge_Canvas_newFromBitmap(bitmap.as_raw_mut()) };
        Canvas { ptr, _marker: PhantomData }
    }

    pub fn delete_owned(self) {
        unsafe { sys::skialin_bridge_Canvas_deleteOwned(self.ptr) };
    }

    pub fn as_raw(&self) -> *mut sys::SkCanvas {
        self.ptr
    }

    fn as_mut(&mut self) -> &mut sys::SkCanvas {
        unsafe { &mut *self.ptr }
    }

    pub fn clear(&mut self, color: Color) {
        unsafe { self.as_mut().clear(color) };
    }

    pub fn draw_color(&mut self, color: Color, mode: BlendMode) {
        unsafe { self.as_mut().drawColor(color, mode.into()) };
    }

    pub fn draw_paint(&mut self, paint: &Paint) {
        unsafe { self.as_mut().drawPaint(&*paint.0) };
    }

    pub fn draw_line(&mut self, p0: Point, p1: Point, paint: &Paint) {
        unsafe { self.as_mut().drawLine1(p0.into(), p1.into(), &*paint.0) };
    }

    pub fn draw_rect(&mut self, rect: Rect, paint: &Paint) {
        let sk_rect: sys::SkRect = rect.into();
        unsafe { self.as_mut().drawRect(&sk_rect, &*paint.0) };
    }

    pub fn draw_oval(&mut self, rect: Rect, paint: &Paint) {
        let sk_rect: sys::SkRect = rect.into();
        unsafe { self.as_mut().drawOval(&sk_rect, &*paint.0) };
    }

    pub fn draw_circle(&mut self, center: Point, radius: f32, paint: &Paint) {
        unsafe { self.as_mut().drawCircle1(center.into(), radius, &*paint.0) };
    }

    pub fn draw_path(&mut self, path: &Path, paint: &Paint) {
        unsafe { self.as_mut().drawPath(path.0, &*paint.0) };
    }

    pub fn draw_text_blob(&mut self, blob: &TextBlob, x: f32, y: f32, paint: &Paint) {
        unsafe { self.as_mut().drawTextBlob(blob.0, x, y, &*paint.0) };
    }

    pub fn save(&mut self) -> i32 {
        unsafe { self.as_mut().save() }
    }

    pub fn restore(&mut self) {
        unsafe { self.as_mut().restore() };
    }

    pub fn restore_to_count(&mut self, save_count: i32) {
        unsafe { self.as_mut().restoreToCount(save_count) };
    }

    pub fn translate(&mut self, dx: f32, dy: f32) {
        unsafe { self.as_mut().translate(dx, dy) };
    }

    pub fn scale(&mut self, sx: f32, sy: f32) {
        unsafe { self.as_mut().scale(sx, sy) };
    }

    pub fn rotate(&mut self, degrees: f32) {
        unsafe { self.as_mut().rotate(degrees) };
    }

    pub fn concat(&mut self, matrix: &Matrix) {
        unsafe { self.as_mut().concat(&matrix.0) };
    }

    pub fn draw_string(&mut self, text: &str, x: f32, y: f32, font: &Font, paint: &Paint) {
        unsafe { self.as_mut().drawSimpleText(text.as_ptr().cast(), text.len(), sys::SkTextEncoding_kUTF8 as sys::SkTextEncoding, x, y, font.0, &*paint.0) };
    }

    pub fn save_count(&self) -> i32 {
        unsafe { sys::SkCanvas_getSaveCount(self.ptr) }
    }

    pub fn rotate_pivot(&mut self, degrees: f32, px: f32, py: f32) {
        unsafe { self.as_mut().rotate1(degrees, px, py) };
    }

    pub fn read_pixels_bitmap(&self, bitmap: &mut Bitmap, src_x: i32, src_y: i32) -> bool {
        unsafe { sys::SkCanvas_readPixels2(self.ptr, bitmap.as_raw_mut(), src_x, src_y) }
    }

    pub fn write_pixels_bitmap(&mut self, bitmap: &Bitmap, x: i32, y: i32) -> bool {
        unsafe { sys::SkCanvas_writePixels1(self.ptr, bitmap.as_raw(), x, y) }
    }

    pub fn draw_picture_with_matrix(&mut self, picture: &Picture, matrix: Option<&Matrix>, paint: Option<&Paint>) {
        let matrix_ptr = matrix.map_or(std::ptr::null(), |m| &m.0 as *const sys::SkMatrix);
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        unsafe { sys::SkCanvas_drawPicture2(self.ptr, picture.0, matrix_ptr, paint_ptr) };
    }

    pub fn draw_drawable_at(&mut self, drawable: &crate::Drawable, x: f32, y: f32) {
        unsafe { self.as_mut().drawDrawable1(drawable.as_raw(), x, y) };
    }

    pub fn local_to_device(&self) -> M44 {
        unsafe { M44::from_raw(sys::skialin_bridge_Canvas_getLocalToDevice(self.ptr)).expect("getLocalToDevice never returns null") }
    }

    pub fn clip_rect(&mut self, rect: Rect, op: ClipOp, antialias: bool) {
        let sk_rect: sys::SkRect = rect.into();
        unsafe { self.as_mut().clipRect(&sk_rect, op.into(), antialias) };
    }

    pub fn clip_path(&mut self, path: &Path, op: ClipOp, antialias: bool) {
        unsafe { self.as_mut().clipPath(path.0, op.into(), antialias) };
    }

    pub fn skew(&mut self, sx: f32, sy: f32) {
        unsafe { self.as_mut().skew(sx, sy) };
    }

    pub fn reset_matrix(&mut self) {
        unsafe { self.as_mut().resetMatrix() };
    }

    pub fn set_matrix(&mut self, matrix: &Matrix) {
        unsafe { self.as_mut().setMatrix1(&matrix.0) };
    }

    pub fn total_matrix(&self) -> Matrix {
        let mut out = Matrix::identity();
        unsafe { sys::skialin_bridge_Canvas_getTotalMatrix(self.ptr, &mut out.0) };
        out
    }

    pub fn quick_reject_rect(&self, rect: Rect) -> bool {
        let sk_rect: sys::SkRect = rect.into();
        unsafe { (*self.ptr).quickReject(&sk_rect) }
    }

    pub fn quick_reject_path(&self, path: &Path) -> bool {
        unsafe { (*self.ptr).quickReject1(path.0) }
    }

    pub fn draw_round_rect(&mut self, rect: Rect, rx: f32, ry: f32, paint: &Paint) {
        let sk_rect: sys::SkRect = rect.into();
        unsafe { self.as_mut().drawRoundRect(&sk_rect, rx, ry, &*paint.0) };
    }

    pub fn draw_arc(&mut self, oval: Rect, start_angle: f32, sweep_angle: f32, use_center: bool, paint: &Paint) {
        let sk_rect: sys::SkRect = oval.into();
        unsafe { self.as_mut().drawArc(&sk_rect, start_angle, sweep_angle, use_center, &*paint.0) };
    }

    pub fn draw_points(&mut self, mode: PointMode, points: &[Point], paint: &Paint) {
        let sk_points: Vec<sys::SkPoint> = points.iter().map(|&p| p.into()).collect();
        let span = sys::SkSpan { _phantom_0: std::marker::PhantomData, fPtr: sk_points.as_ptr().cast_mut(), fSize: sk_points.len() };
        unsafe { self.as_mut().drawPoints(mode.into(), span, &*paint.0) };
    }

    pub fn draw_image(&mut self, image: &Image, x: f32, y: f32, sampling: SamplingOptions, paint: Option<&Paint>) {
        let sk_sampling = to_sk_sampling(sampling);
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        unsafe { self.as_mut().drawImage2(image.0, x, y, &sk_sampling, paint_ptr) };
    }

    #[allow(clippy::too_many_arguments)]
    pub fn draw_image_rect(&mut self, image: &Image, src: Option<Rect>, dst: Rect, sampling: SamplingOptions, paint: Option<&Paint>, constraint: SrcRectConstraint) {
        let sk_dst: sys::SkRect = dst.into();
        let sk_sampling = to_sk_sampling(sampling);
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        match src {
            // `src` is a required C++ reference in this overload, so it can never be passed as null.
            Some(src) => {
                let sk_src: sys::SkRect = src.into();
                unsafe { self.as_mut().drawImageRect(image.0, &sk_src, &sk_dst, &sk_sampling, paint_ptr, constraint.into()) };
            }
            None => unsafe { self.as_mut().drawImageRect1(image.0, &sk_dst, &sk_sampling, paint_ptr) },
        }
    }

    pub fn draw_image_nine(&mut self, image: &Image, center: IRect, dst: Rect, filter: FilterMode, paint: Option<&Paint>) {
        let sk_center: sys::SkIRect = center.into();
        let sk_dst: sys::SkRect = dst.into();
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        unsafe { self.as_mut().drawImageNine(image.0, &sk_center, &sk_dst, filter.into(), paint_ptr) };
    }

    pub fn save_layer(&mut self, bounds: Option<Rect>, paint: Option<&Paint>) -> i32 {
        let sk_bounds: Option<sys::SkRect> = bounds.map(Into::into);
        let bounds_ptr = sk_bounds.as_ref().map_or(std::ptr::null(), |r| r as *const sys::SkRect);
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        unsafe { self.as_mut().saveLayer(bounds_ptr, paint_ptr) }
    }

    pub fn save_layer_with_backdrop(&mut self, bounds: Option<Rect>, paint: Option<&Paint>, backdrop: Option<&crate::ImageFilter>, flags: u32) -> i32 {
        let sk_bounds: Option<sys::SkRect> = bounds.map(Into::into);
        let bounds_ptr = sk_bounds.as_ref().map_or(std::ptr::null(), |r| r as *const sys::SkRect);
        let paint_ptr = paint.map_or(std::ptr::null(), |p| &*p.0 as *const sys::SkPaint);
        let backdrop_ptr = backdrop.map_or(std::ptr::null_mut(), |f| f.0);
        unsafe { sys::skialin_bridge_Canvas_saveLayer(self.ptr, bounds_ptr, paint_ptr, backdrop_ptr, flags) }
    }

    pub fn draw_rrect(&mut self, rrect: &RRect, paint: &Paint) {
        unsafe { sys::skialin_bridge_Canvas_drawRRect(self.ptr, rrect.0, &*paint.0) };
    }

    pub fn draw_picture(&mut self, picture: &crate::Picture) {
        unsafe { sys::skialin_bridge_Canvas_drawPicture(self.ptr, picture.0) };
    }

    pub fn draw_drrect(&mut self, outer: &RRect, inner: &RRect, paint: &Paint) {
        unsafe { sys::skialin_bridge_Canvas_drawDRRect(self.ptr, outer.0, inner.0, &*paint.0) };
    }

    pub fn clip_rrect(&mut self, rrect: &RRect, op: ClipOp, antialias: bool) {
        unsafe { sys::skialin_bridge_Canvas_clipRRect(self.ptr, rrect.0, op.into(), antialias) };
    }

    pub fn draw_region(&mut self, region: &Region, paint: &Paint) {
        unsafe { self.as_mut().drawRegion(region.0, &*paint.0) };
    }

    pub fn clip_region(&mut self, region: &Region, op: ClipOp) {
        unsafe { self.as_mut().clipRegion(region.0, op.into()) };
    }

    pub fn draw_vertices(&mut self, vertices: &Vertices, mode: BlendMode, paint: &Paint) {
        unsafe { sys::skialin_bridge_Canvas_drawVertices(self.ptr, vertices.0, mode.into(), &*paint.0) };
    }

    pub fn draw_patch(&mut self, cubics: [Point; 12], colors: [Color; 4], tex_coords: Option<[Point; 4]>, mode: BlendMode, paint: &Paint) {
        let sk_cubics: Vec<sys::SkPoint> = cubics.iter().map(|&p| p.into()).collect();
        let sk_tex: Option<[sys::SkPoint; 4]> = tex_coords.map(|coords| std::array::from_fn(|i| coords[i].into()));
        let tex_ptr = sk_tex.as_ref().map_or(std::ptr::null(), |t| t.as_ptr());
        unsafe { self.as_mut().drawPatch(sk_cubics.as_ptr(), colors.as_ptr(), tex_ptr, mode.into(), &*paint.0) };
    }

    pub fn draw_annotation(&mut self, rect: Rect, key: &str, value: Option<&Data>) {
        let sk_rect: sys::SkRect = rect.into();
        let key_cstr = std::ffi::CString::new(key).expect("annotation key must not contain a NUL byte");
        let value_ptr = value.map_or(std::ptr::null_mut(), |d| d.0);
        unsafe { self.as_mut().drawAnnotation(&sk_rect, key_cstr.as_ptr(), value_ptr) };
    }

    pub fn draw_drawable(&mut self, drawable: &crate::Drawable, matrix: Option<&Matrix>) {
        let matrix_ptr = matrix.map_or(std::ptr::null(), |m| &m.0 as *const sys::SkMatrix);
        unsafe { self.as_mut().drawDrawable(drawable.as_raw(), matrix_ptr) };
    }

    pub fn concat_44(&mut self, matrix: &M44) {
        unsafe { sys::skialin_bridge_Canvas_concat44(self.ptr, matrix.0) };
    }

    #[allow(clippy::too_many_arguments)]
    pub fn draw_shadow(
        &mut self,
        path: &Path,
        z_plane: (f32, f32, f32),
        light_pos: (f32, f32, f32),
        light_radius: f32,
        ambient_color: Color,
        spot_color: Color,
        flags: u32,
    ) {
        unsafe {
            sys::skialin_bridge_ShadowUtils_drawShadow(
                self.ptr,
                path.0,
                z_plane.0,
                z_plane.1,
                z_plane.2,
                light_pos.0,
                light_pos.1,
                light_pos.2,
                light_radius,
                ambient_color,
                spot_color,
                flags,
            )
        };
    }
}
